### PR TITLE
feat(CLOUDDST-31926): add duplicate package validation for FBC fragments

### PR DESCRIPTION
**Pull request by yashvardhannanavati**
Implements validation to detect duplicate packages across FBC fragments targeting the same OCP version, preventing conflicts during catalog publishing of binaryless FBC fragments. This change does not hinder publishing of binary-enabled FBCs in any way.

  Changes:
  - Add duplicate package detection in prepare-fbc-parameters task
    - Track OCP version to component mapping during fragment processing
    - Store extracted packages for each component from opm render output
    - Validate no duplicate packages exist within same OCP version group
    - Report detailed errors with component names when duplicates found

  - Update test mocks to handle multiple test scenarios
    - Support both original and operator-foundry bundle images
    - Dynamically return opt-in results based on requested bundles

  - Add test-prepare-fbc-parameters-duplicate-packages.yaml
    - Verifies duplicate package validation fails correctly
    - Tests 4 components (2 pairs of duplicates) targeting same OCP version
    - Uses test/assert-task-failure annotation for expected failure

Assisted-By: Claude

### CODE REVIEW

File: tasks/managed/prepare-fbc-parameters/tests/mocks.sh [hunk 1]
```diff
@@ -15,12 +15,25 @@ function kubectl() {
...
   elif [[ "$*" == "get internalrequest test-check-fbc-opt-in-ir -o jsonpath={.status.results}" ]]; then
-    # Check if this is the failure propagation test based on the allowedPackages in data.json
-    if [ -f "$(params.dataDir)/data.json" ] && grep -q "non-matching-package" "$(params.dataDir)/data.json"; then
-      # Return mock opt-in results for failure propagation test (all opted in, but packages will fail validation)
+    # Read what bundle images were actually requested from the internal-request call
+    if [ -f "$(params.dataDir)/mock_internal-request.txt" ]; then
+      requested_bundles=$(sed -n 's/.*containerImages=\(\[.*\]\).*/\1/p' \
+        "$(params.dataDir)/mock_internal-request.txt" || echo "")
+    else
+      requested_bundles=""
```
Comment (theflockers) [inline tasks/managed/prepare-fbc-parameters/tests/mocks.sh]: maybe setting `requested_bundles=""` before the if, so you don't need an `else` block as it will only be modified in case the next conditional is true.

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml [hunk 2]
```diff
@@ -305,11 +313,78 @@ spec:
...
+                continue
+            fi
+
+            echo "  INFO: OCP version ${ocp_ver}: ${component_count_for_version} components -" \
+              "checking for duplicate packages..."
+
+            # Build package -> component mapping for this OCP version
+            declare -A package_to_component
+
+            while IFS= read -r comp_idx; do
```
Comment (theflockers) [inline tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml:360]: In release we normally avoid using while loops changing IFS as it _eventually_ might not be deterministic depending on the input. If some string comes weirdly formatted, or with extra spaces (or whatever) it tends to  behave differently.  The second reason is because `while` loops creates a subshell (which is fine in this case as you don't need to set IFS back).
So we _advice_ to check the size of `comp_idx` and use a `for` loop, unless there is no risk of weird input to `$comp_index` or `$pkg`

Comment (theflockers) [reply tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml:360]: A late comment. I am considering that you are using a `json` string as input. So that is why I am saying that.

Comment (yashvardhannanavati) [reply tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml:360]: I did have for loops and qodo suggested changing them to while here: https://github.com/konflux-ci/release-service-catalog/pull/2033#discussion_r2868386463

Comment (yashvardhannanavati) [reply tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml:360]: It's not json input. It's simply new-line delimited data

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml [hunk 3]
```diff
@@ -238,13 +243,16 @@ spec:
             echo "  INFO: Validating packages in fragment..."
             rm -f "${RENDERED_CATALOG}"
             opm render "${fbc_fragment}" > "${RENDERED_CATALOG}"
-            actual_packages=$(jq -r 'select(.schema=="olm.package") | .name' < "${RENDERED_CATALOG}" | uniq)
+            actual_packages=$(jq -r 'select(.schema=="olm.package") | .name' < "${RENDERED_CATALOG}" | sort -u)
             component_bundle_images=$(jq -r \
               'select(.schema == "olm.bundle") | "\(.image)" | split("@")[0]' \
               < "${RENDERED_CATALOG}" | uniq | jq -R . | jq -s .)
```
Comment (elenagerman) [inline tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml]: minor, but probably it should be changed from 'uniq' to 'sort -u' here as well

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml [hunk 4]
```diff
@@ -305,11 +313,78 @@ spec:
...
+                    fi
+                done <<< "$packages"
+            done <<< "$component_indices"
+
+            # Clean up for next iteration
+            unset package_to_component
+
+            if [ "$validation_failed" != "true" ]; then
+                echo "    SUCCESS: No duplicate packages found for OCP version ${ocp_ver}"
+            fi
```
Comment (elenagerman) [inline tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml:390]: It is worth adding a local variable that resets at the top of each loop iteration, and using that for the success message check instead of the global flag (validation_failed) to control the log message.
Otherwise, the logs could look misleading: once any ocp version fails, the global flag stays true, so every subsequent version, even clean ones, silently skips printing its success message.

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml (hunk 2 above)
Comment (theflockers) [reply tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml:360]: ok. resolving.

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml (hunk 4 above)
Comment (yashvardhannanavati) [reply tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml:390]: good point, fixed.

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml (hunk 3 above)
Comment (yashvardhannanavati) [reply tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml]: yes, thanks for catching that. fixed

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml [hunk 5]
```diff
@@ -305,11 +313,78 @@ spec:
...
+        done
```
[the remark this answers is not in the document]
Comment (yashvardhannanavati) [reply tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml:391]: done

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml [hunk 6]
```diff
@@ -305,11 +313,78 @@ spec:
...
+        # causing data loss. This validation enforces package uniqueness within the scope of each
+        # target OCP version, while allowing the same package to exist across different versions.
+        echo "INFO: Validating no duplicate packages within FBCs for the same OCP version..."
+
+        # For each OCP version that has multiple components
+        for ocp_ver in "${!ocp_version_to_component_indices[@]}"; do
+            component_indices="${ocp_version_to_component_indices[$ocp_ver]}"
+
+            # Count components for this OCP version (count lines instead of words)
+            component_count_for_version=$(echo -n "$component_indices" | grep -c '^' || echo 1)
```
[the remark this answers is not in the document]
Comment (yashvardhannanavati) [reply tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml]: done

**Review by elenagerman — DISMISSED** [review]

File: tasks/managed/prepare-fbc-parameters/tests/mocks.sh (hunk 1 above)
Comment (yashvardhannanavati) [reply tasks/managed/prepare-fbc-parameters/tests/mocks.sh]: done, fixed.

**Review by theflockers — DISMISSED** [review]

**Review by davidmogar — DISMISSED** [review]

**Review by nmars — APPROVED** [review]
LGTM. My first review of a PR in the release service. I really like the way the updated task was run in [Run Tekton Task tests](https://github.com/konflux-ci/release-service-catalog/actions/runs/22891899186/job/66416768117?pr=2033#logs).

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml [hunk 7]
```diff
@@ -305,11 +313,82 @@ spec:
...
+            declare -A package_to_component
+
+            while IFS= read -r comp_idx; do
+                packages="${component_to_packages[$comp_idx]}"
+                while IFS= read -r pkg; do
+                    [ -z "$pkg" ] && continue  # Skip empty lines
+                    if [ -n "${package_to_component[$pkg]:-}" ]; then
+                        # Duplicate found!
+                        first_component_idx="${package_to_component[$pkg]}"
+                        first_component_name=$(jq -cr --argjson idx "$first_component_idx" \
```
Comment (johnbieren) [inline tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml]: nit - the `-c` doesn't do anything here (or line 368)

**Review by johnbieren — DISMISSED** [review]

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml [hunk 8]
```diff
@@ -305,11 +313,82 @@ spec:
...
+        # causing data loss. This validation enforces package uniqueness within the scope of each
+        # target OCP version, while allowing the same package to exist across different versions.
+        echo "INFO: Validating no duplicate packages within FBCs for the same OCP version..."
+
+        # For each OCP version that has multiple components
+        for ocp_ver in "${!ocp_version_to_component_indices[@]}"; do
+            component_indices="${ocp_version_to_component_indices[$ocp_ver]}"
+
+            # Count components for this OCP version (count lines, filtering empty lines)
+            component_count_for_version=$(printf '%s\n' "$component_indices" | sed '/^$/d' | wc -l | tr -d ' ')
```
Comment (lipoja) [inline tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml]: Can it happen that we have also line containing just whitespace? 
I am asking just because I saw the `sed` command and I was wondering if `sed '/^[[:space:]]*$/d'` would not be more robust. However if we are sure it will not happen then we can keep it simple.

**Review by lipoja — APPROVED** [review]

**Comment by davidmogar** [comment]
@yashvardhannanavati tests for fbc are not the best right now. Could you add some comments about what is this missing, where the issue was identified and whether the tests in this PR are enough to ensure no errors will be introduced?

**Comment by yashvardhannanavati** [comment]
@davidmogar I do have a comment in the PR explaining changes in detail here: https://github.com/konflux-ci/release-service-catalog/pull/2033#issue-4004065178

This changes in this PR will not affect the existing workflow. New tests have been added to cover success and failure scenarios.

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml (hunk 7 above)
Comment (yashvardhannanavati) [reply tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml]: good catch! addressed it. Could you please help re-approve the MR?

File: tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml (hunk 8 above)
Comment (yashvardhannanavati) [reply tasks/managed/prepare-fbc-parameters/prepare-fbc-parameters.yaml]: well, the data comes from lines 315-318 so under normal circumstances, we shouldn't get a line with just whitespace. But I like the defensive coding option. Changed it. Nice catch!

**Comment by yashvardhannanavati** [comment]
@theflockers @elenagerman @davidmogar @johnbieren  fixing nits wiped out the approvals. Could you please help re-approve this?

**Review by johnbieren — APPROVED** [review]

**Review by davidmogar — APPROVED** [review]

**Review by gbenhaim — APPROVED** [review]
